src/java/org/apache/cassandra/gms/Gossiper.java

```diff
@@ -136,6 +139,38 @@
 
     private volatile long lastProcessedMessageAt = System.currentTimeMillis();
 
+    private boolean haveMajorVersion3Nodes = true;
+
+    final com.google.common.base.Supplier<Boolean> haveMajorVersion3NodesSupplier = () ->
+    {
+        //Once there are no prior version nodes we don't need to keep rechecking
+        if (!haveMajorVersion3Nodes)
+            return false;
+
+        Iterable<InetAddressAndPort> allHosts = Iterables.concat(Gossiper.instance.getLiveMembers(), Gossiper.instance.getUnreachableMembers());
+        CassandraVersion referenceVersion = null;
+
+        for (InetAddressAndPort host : allHosts)
+        {
+            CassandraVersion version = getReleaseVersion(host);
+
+            //Raced with changes to gossip state
+            if (version == null)
+                continue;
+
+            if (referenceVersion == null)
+                referenceVersion = version;
+
+            if (version.major < 4)
```

Comment:
This will capture nodes with major version 2 as well. If that is ok, we should probably rename variables to reflect that.

Reply:
> This will capture nodes with major version 2 as well. If that is ok, we should probably rename variables to reflect that.

Nvm this. We discussed offline, version 2 nodes cannot connection to version 4 nodes. Enforced by 
https://github.com/apache/cassandra/blob/trunk/src/java/org/apache/cassandra/net/async/InboundHandshakeHandler.java#L180